drivers/ethernet/eth_xilinx_axienet.c

```diff
@@ -0,0 +1,577 @@
+#define XILINX_AXIENET_INTERRUPT_PENDING_MGTRDY_MASK      0x00000080 /* MGT clock Lock */
+#define XILINX_AXIENET_INTERRUPT_PENDING_PHYRSTCMPLT_MASK 0x00000100 /* Phy Reset complete */
+
+#define XILINX_AXIENET_INTERRUPT_PENDING_RXREJ_MASK 0x00000008
+#define XILINX_AXIENET_INTERRUPT_ENABLE_OFFSET      0x00000014
+#define XILINX_AXIENET_INTERRUPT_ENABLE_RXREJ_MASK  0x00000008
+#define XILINX_AXIENET_INTERRUPT_ENABLE_OVR_MASK    0x00000010 /* FIFO overrun */
+
+#define XILINX_AXIENET_RECEIVER_CONFIGURATION_WORD_0_REG_OFFSET     0x00000400
+#define XILINX_AXIENET_RECEIVER_CONFIGURATION_WORD_1_REG_OFFSET     0x00000404
+#define XILINX_AXIENET_RECEIVER_CONFIGURATION_WORD_1_REG_RX_EN_MASK 0x10000000
+#define XILINX_AXIENET_RECEIVER_CONFIGURATION_FLOW_CONTROL_OFFSET   0x0000040C
+#define XILINX_AXIENET_RECEIVER_CONFIGURATION_FLOW_CONTROL_EN_MASK  0x20000000
+#define XILINX_AXIENET_TX_CONTROL_REG_OFFSET                        0x00000408
+#define XILINX_AXIENET_TX_CONTROL_TX_EN_MASK                        (1 << 11)
+
+#define XILINX_AXIENET_UNICAST_ADDRESS_WORD_0_OFFSET 0x00000700
+#define XILINX_AXIENET_UNICAST_ADDRESS_WORD_1_OFFSET 0x00000704
+
+#define ETH_ALEN 6
+
+/* device state */
+struct xilinx_axienet_data {
+	/* device mac address */
+	uint8_t mac_addr[ETH_ALEN];
+	uint8_t tx_buffer[NET_ETH_MTU];
+	uint8_t rx_buffer[NET_ETH_MTU];
+
+	struct net_if *interface;
+
+	bool dma_is_configured_rx, dma_is_configured_tx;
+};
+
+/* global configuration per Ethernet device */
+struct xilinx_axienet_config {
+	void (*config_func)(const struct xilinx_axienet_data *dev);
+	const struct device *dma;
+
+	const struct device *phy;
+
+	void *reg;
+
+	bool have_irq, have_rx_csum_offload, have_tx_csum_offload;
+};
+
+static void xilinx_axienet_write_register(const struct xilinx_axienet_config *config,
+					  uintptr_t reg_offset, uint32_t value)
+{
+	volatile uint32_t *reg_addr = (uint32_t *)((uint8_t *)(config->reg) + reg_offset);
+
+	*reg_addr = value;
+	barrier_dmem_fence_full(); /* make sure that write commits */
+}
+
+static uint32_t xilinx_axienet_read_register(const struct xilinx_axienet_config *config,
+					     uintptr_t reg_offset)
+{
+	const volatile uint32_t *reg_addr = (uint32_t *)((uint8_t *)(config->reg) + reg_offset);
+	const uint32_t ret = *reg_addr;
+
+	barrier_dmem_fence_full(); /* make sure that read commits */
+	return ret;
+}
+static int setup_dma_rx_transfer(const struct device *dev,
+				 const struct xilinx_axienet_config *config,
+				 struct xilinx_axienet_data *data);
+
+/* called by DMA when a packet is available */
+static void xilinx_axienet_rx_callback(const struct device *dma, void *user_data, uint32_t channel,
+				       int status)
+{
+	struct device *ethdev = (struct device *)user_data;
+	struct xilinx_axienet_data *data = ethdev->data;
+
+	(void)ethdev;
+	if (status < 0) {
+		LOG_ERR("DMA RX error: %d", status);
+		k_panic();
+		eth_stats_update_errors_rx(data->interface);
+	} else {
```

Comment:
you can do a return from the if block so that you dont need an else block. Generally zephyr follows linux coding guidelines which encourages limited amounts of indents and nested blocks.

Reply:
In this case, I would have to duplicate the call to setup_dma_rx_transfer. I need to setup a new RX transfer regardless of whether the DMA reports an error, otherwise, the ethernet subsystem just stops receiving.

Reply:
Actually, nevermind - I have replaced the nested if-then-else's with a goto.